src/vos/ilog.c

```diff
@@ -1552,3 +1392,257 @@ ilog_fetch_finish(struct ilog_entries *entries)
+		}
+		if (entry->ie_punch == punch) {
+			/* Remove redundant entry */
+			D_GOTO(done, rc = AGG_RC_REMOVE);
+		}
+	}
+
+	if (!entry->ie_punch) {
+		/* Create is needed for now */
+		D_GOTO(done, rc = AGG_RC_NEXT);
+	}
+
+	if (agg_arg->aa_prev == NULL) {
+		/* No punched entry to remove */
+		D_GOTO(done, rc = AGG_RC_REMOVE);
+	}
+
+	if (agg_arg->aa_prev->ie_id.id_epoch < agg_arg->aa_epr->epr_lo) {
+		/** Data punched is not in range */
+		agg_arg->aa_prior_punch = entry;
+		D_GOTO(done, rc = AGG_RC_NEXT);
+	}
+
+	D_ASSERT(!agg_arg->aa_prev->ie_punch);
+	/* Punch is redundant or covers nothing.  Remove it. */
+	rc = AGG_RC_REMOVE_PREV;
+done:
+	return rc;
+}
+
+int
+ilog_aggregate(struct umem_instance *umm, struct ilog_df *ilog,
+	       const struct ilog_desc_cbs *cbs, const daos_epoch_range_t *epr,
+	       bool discard, daos_epoch_t punched, struct ilog_entries *entries)
+{
+	struct ilog_priv	*priv = ilog_ent2priv(entries);
+	struct ilog_context	*lctx;
+	struct ilog_entry	*entry;
+	struct agg_arg		 agg_arg;
+	struct ilog_root	*root;
+	struct ilog_root	 tmp = {0};
+	struct ilog_id		 old_id;
+	struct umem_attr	 uma;
+	bool			 empty = false;
+	int			 rc = 0;
+	int			 removed = 0;
+	daos_handle_t		 toh = DAOS_HDL_INVAL;
+
+	D_ASSERT(epr != NULL);
+	D_ASSERT(punched <= epr->epr_hi);
+
+	D_DEBUG(DB_IO, "%s incarnation log: epr: "DF_U64"-"DF_U64" punched="
+		DF_U64"\n", discard ? "Discard" : "Aggregate", epr->epr_lo,
+		epr->epr_hi, punched);
+
+	/* This can potentially be optimized but using ilog_fetch gets some code
+	 * reuse.
+	 */
+	rc = ilog_fetch(umm, ilog, cbs, DAOS_INTENT_PURGE, entries);
+	if (rc == -DER_NONEXIST) {
+		D_DEBUG(DB_IO, "log is empty\n");
+		/* Log is empty */
+		return 1;
+	}
+
+	lctx = &priv->ip_lctx;
+
+	root = lctx->ic_root;
+
+	ILOG_ASSERT_VALID(root);
+
+	D_ASSERT(!ilog_empty(root)); /* ilog_fetch should have failed */
+
+	agg_arg.aa_epr = epr;
+	agg_arg.aa_prev = NULL;
+	agg_arg.aa_prior_punch = NULL;
+	agg_arg.aa_punched = punched;
+	agg_arg.aa_discard = discard;
+
+	if (root->lr_tree.it_embedded) {
+		entry = &entries->ie_entries[0];
+		rc = check_agg_entry(&entries->ie_entries[0], &agg_arg);
+
+		switch (rc) {
+		case AGG_RC_DONE:
+		case AGG_RC_NEXT:
+			rc = 0;
+			break;
+		case AGG_RC_REMOVE:
+			old_id = root->lr_id;
+			tmp.lr_magic = ilog_ver_inc(lctx);
+			rc = ilog_ptr_set(lctx, root, &tmp);
+			if (rc != 0)
+				break;
+
+			empty = true;
+			rc = ilog_log_del(lctx, &old_id);
+			D_DEBUG(DB_IO, "Removed ilog entry at "DF_U64" "DF_RC
+				"\n", entry->ie_id.id_epoch, DP_RC(rc));
+			if (rc == 0)
+				removed++;
+			break;
+		case AGG_RC_REMOVE_PREV:
+			/* Fall through: Should not get this here */
+		default:
+			D_ASSERT(0);
+		}
+		goto done;
+	}
+
+	umem_attr_get(&lctx->ic_umm, &uma);
+	rc = dbtree_open(root->lr_tree.it_root, &uma, &toh);
+	if (rc != 0) {
+		D_ERROR("Failed to open incarnation log tree: "DF_RC
+			"\n", DP_RC(rc));
+		return rc;
+	}
+	ilog_foreach_entry(entries, entry) {
```

Comment:
The epoch for the entries in the array are ordered or not?

Reply:
Yes